runtime/lua/vim/lsp.lua

```diff
@@ -890,6 +890,42 @@ function lsp.start_client(config)
     end)
   end
 
+  --@private
+  --- Sends a request to the server and synchronously waits for the response.
+  ---
+  --- This is a wrapper around {client.request}
+  ---
+  --@param method (string) LSP method name.
+  --@param params (table) LSP request params.
+  --@param timeout_ms (number, optional, default=100) Maximum time in
+  ---milliseconds to wait for a result.
+  --@param bufnr (number) Buffer handle (0 for current).
+  --@returns { err, result }, where `err` and `result` come from the |lsp-handler|.
+  ---On timeout, cancel or error, returns `(nil, err)` where `err` is a
+  ---string describing the failure reason. If the request was unsuccessful
+  ---returns `nil`.
+  --@see |vim.lsp.buf_request_sync()|
```

Comment:
The docs in `vim.lsp.client` should probably be extended so that this method shows up as well.

That there are 3 error scenarios also bothers me a bit, but I can see that it could be useful to distinguish between all of them, so I guess it's okay that way.

Reply:
Yes it's a bit weird. I did this to stay as close as possible to `buf_request_sync` which also has those 3 error scenarios; when none of the requests are successful it returns `nil` (and until recently it would return `{}`).

I interpret this as: returns `nil` when something went wrong, and returns an optional second value `err` with details about the error.

Reply:
Added doc in `vim.lsp.client`